apollo-router-core/src/query_planner/mod.rs

```diff
@@ -17,7 +17,7 @@ pub struct QueryPlanOptions {}
 
 #[derive(Debug)]
 pub struct QueryPlan {
-    root: PlanNode,
+    root: Option<PlanNode>,
```

Comment:
:screwdriver: Keep this one not optional and make that one optional instead and handle the exception in that file too: https://github.com/apollographql/router/blob/c759f464e294fbdd8c08fa94257bba21ed016a66/apollo-router-core/src/query_planner/router_bridge_query_planner.rs#L67

This is because either we have a plan, either it failed to give us a plan. Having "maybe a plan" makes no sense.

This will remove the panic and allow you to handle the error instead of dealing with options everywhere.

```suggestion
    root: PlanNode,
```

Reply:
Hmm so an empty QueryPlan would be a QueryPlan with an empty root instead of a QueryPlan with no root?

IMO an empty plan is not a failure, it just means there's nothing to do.

Reply:
> IMO an empty plan is not a failure, it just means there's nothing to do.

Yes well. That is really not clear imo.

Originally, router did handle the optional plan node:

https://github.com/apollographql/router/blob/31eb0d8e290676d46b41382cb06780093b84b33e/apollo-router/src/apollo_router.rs#L95-L101

I removed this if-else in #233 and clearly nobody reacted to this change so I suspect it was not clear for anybody why this was an Option in the first place.

> - Parse planner result as a PlannerResult instead of QueryPlanner: this new PlannerResult enum can detect only the kind QueryPlanner and will return a specific error on parsing if something else is returned by the router bridge.

I had no idea why this thing was an option but it wasn't failing any test and there was no comment or no clue to tell me why it was like that. So I guessed it happened when the query planer failed to make a plan. I was half-right.

So now I understand that the query planner can return no plan if there is nothing to do. If we want to make this more explicit, then we should keep the `QueryPlanner` having a required `PlanNode` and handle the "no plan scenario" as an explicit error variant in `QueryPlannerError` that says "no plan because this is naive introspection" (or something like that).

I think it's best to be explicit here using a descriptive error rather than thinking "oh everybody knows that if no plan is returned it means the introspection query wasn't handled". The user of the API will thank you for that.

Also: what needs to happen exactly when this naive introspection query was unknown? Is there a specific response for that? Something in errors? This should be explicit in the code. In the past the behavior was to return an empty stream so no response at all. I suspect it should be logged and the end user should have a response with a proper error.

Reply:
Totally agree, I'm going to try to see if we have a way to "check for a naive introspection query beforehand", and at least return something like "sorry but we don't support this yet" or something.

Reply:
fixed in [a348c09](https://github.com/apollographql/router/pull/281/commits/a348c0932356b0f4c96b51be9eb73c7ecb6a2ac7)